tidalsim/bb/spike.py

```diff
@@ -41,50 +42,35 @@ def spike_trace_to_bbs(trace: Iterator[SpikeTraceEntry]) -> BasicBlocks:
     # No data is stored, speeds up the lookup significantly
     start = None
     previous_inst: Optional[SpikeTraceEntry] = None
-    intervals = IntervalTree()
+    intervals: List[Tuple[int, int]] = []
     for trace_entry in tqdm(trace):
         if start is None:
             start = trace_entry.pc
         if trace_entry.decoded_inst in control_insts:
-            intervals[start:trace_entry.pc + 1] = None # Intervals are inclusive of the start, but exclusive of the end
+            intervals += [(start, trace_entry.pc+1)] # Intervals are inclusive of the start, but exclusive of the end
             start = None
         if previous_inst and (abs(trace_entry.pc - previous_inst.pc) > 4) and previous_inst.decoded_inst not in control_insts:
             raise RuntimeError(f"Control diverged from PC: {hex(previous_inst.pc)} \
                     to PC: {hex(trace_entry.pc)}, but the last instruction {previous_inst.decoded_inst} \
                     wasn't a control instruction")
         previous_inst = trace_entry
-    if start is not None and previous_inst is not None:
-        intervals[start:previous_inst.pc+1] = None
 
-    # Ensures that PCs have a one-to-one mapping to an interval
-    intervals.merge_equals()
-    intervals.split_overlaps()
+    if start is not None and previous_inst is not None:
+       intervals += [(start, previous_inst.pc+1)]
 
-    # Assign each interval an index in the BBV vector
-    # Easier to just instantiate a new tree for lookup later
-    id = 0
-    unique_intervals = IntervalTree()
-    for interval in sorted(intervals.items()):
-        unique_intervals.addi(interval.begin, interval.end, id)
-        id += 1
-    return BasicBlocks(pc_to_bb_id=unique_intervals)
+    return BasicBlocks(markers=intervals_to_markers(intervals))
 
 def spike_trace_to_embedding_df(trace: Iterator[SpikeTraceEntry], bb: BasicBlocks, interval_length: int) -> DataFrame[EmbeddingSchema]:
     # Dimensions of dataframe
     # # rows = # of intervals = ceil( (length of trace) / interval_length )
     # # cols = # of features = # of elements in the intervaltree
-    n_features = len(bb.pc_to_bb_id)
-
-    # Use a cache to avoid querying the interval tree too often, queries should have good locality
-    @functools.lru_cache(maxsize=128)
```

Comment:
Could you leave this cache in place? In my experiment, I see a throughput of 480 intervals/second without the cache and ~700 intervals/second with the cache. So this seems marginally faster than the intervaltree solution, but it is much cleaner imo and avoids a dependency :)

Reply:
that seems very slow to me 🤔. What were the original benchmarks and what does intervals/second mean here?

Reply:
The benchmark used is the aha-mont64 binary from embench. The interval length is 1000 instructions and intervals/second means how many such intervals we were able to compute the embedding for each second by processing the spike commit log. So it basically means the current implementation can process ~700k RISC-V instructions / second when performing embedding.